packages/gatsby-theme-newrelic/src/components/SearchInput.js

```diff
@@ -1,82 +1,103 @@
-import React from 'react';
+import React, { forwardRef } from 'react';
 import PropTypes from 'prop-types';
 import { css } from '@emotion/core';
 import styled from '@emotion/styled';
 import Icon from './Icon';
 
 const SIZES = {
+  SMALL: 'small',
   MEDIUM: 'medium',
   LARGE: 'large',
 };
 
-const SearchInput = ({
-  onClear,
-  value,
-  width,
-  size = 'medium',
-  className,
-  style,
-  ...props
-}) => {
-  const handleClick = (e) => {
-    e.preventDefault();
-    onClear();
-  };
-  const handleKeyDown = (e) => {
-    if (e.key === 'Escape' && onClear) onClear();
-  };
+const SearchInput = forwardRef(
```

Comment:
Would you be able to give me an overview of what the forwardRef is doing? I'm looking at the documentation and currently am not getting it 😅

Reply:
Yes!

So normally when you write components, passing a `ref` to it gives you access to the _instance_ of that component. For example:

```js
class MyComponent extends React.Component {
  launch() {
    alert('Launched!!')
  }

  render() {
    return <div>My component</div>
  }
}

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.launch() // will call the `launch` function on the `MyComponent` instance
  }, [])

  return <MyComponent ref={ref} />
}
```

Notice here how I've used `ref` as "prop" passed to `MyComponent`. This is great if you actually want to interact with the _instance_ of the component and not its underlying DOM node. Let's take another example that you see a lot:

```js
const CustomInput extends React.Component {
  render() {
    return <input {...this.props} className="custom-input" />
  }
}

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.focus()
  }, [])

  return <CustomInput ref={ref} />
}
```

This will blow up. Why? Because the _instance_ of `CustomInput` does not have a `focus` method.

Even worse is that now that functional components are the common thing, there is no way to use a `ref` with a functional component:

```js
const CustomInput = (props) => <input {...props} className="custom-input" />

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.focus()
  }, [])

  // this will blow up because React does not allow you to add a ref on a functional component
  return <CustomInput ref={ref} /> 
}
```

So how might we fix this?

1 way would be to create a custom prop that allowed you to forward that `ref` to the underlying node. Typically you'd see this called `innerRef`:

```js
const CustomInput = ({ innerRef, ...props }) => <input ref={innerRef} {...props} className="custom-input" />

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.focus() // works! innerRef is now attached to the underlying <input /> element
  }, [])

  return <CustomInput innerRef={ref} /> 
}
```

This works well, but notice that it forces me to learn a new concept: I need a new name for the `ref` just so that I can "forward" it to the underlying element. Now I have to be aware of which components use `ref` and which ones expose an `innerRef`.

Another way to fix this would be to rewrite the `CustomInput` to a class component and add an instance method that would forward to the element:

```js
const CustomInput extends React.Component {
  inputRef = createRef()
  
  focus() {
    this.inputRef.current.focus()
  }

  render() {
    return <input ref={this.inputRef} {...this.props} className="custom-input" />
  }
}

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.focus() // works! ref calls the instance method `focus` on the class, which delegates to its own ref on the input
  }, [])

  return <CustomInput ref={ref} /> 
}
```

From a consumer standpoint, this is nice, because I can just use `ref` without having to worry about its implementation. I don't have to learn a new name just to forward the ref to the underlying element. On the flip-side, this sucks as a component author because I really like functional components and I want to continue using them. It would be awful to have to rewrite my components to class components, or have to create a custom prop just so that I can forward the ref to the underlying element.

This is where `forwardRef` comes in. It wraps function components so that you can forward refs to underlying elements, but still allow consumers to use the `ref` name. This is a much cleaner approach. Now the implementation looks like this:

```js
const CustomInput = forwardRef((props, ref) => <input ref={ref} {...props} className="custom-input" />)

const AnotherComponent = () => {
  const ref = useRef()

  useEffect(() => {
    ref.current.focus() // works! ref has been forwarded by the CustomInput to the underlying <input /> element 
  }, [])

  return <CustomInput ref={ref} /> 
}
```

As you can see, this is MUCH cleaner and avoids the tradeoffs you have to make above. As a consumer of the component, you only have to know 1 concept (`ref`) and as an author, you don't have to sacrifice the way you like to author components just to forward refs.

Hopefully this helps!

Reply:
i was wondering the same. thanks for explaining that!

Reply:
yes! thank you so much for this explanation, that makes a lot of sense rather than passing the ref down continuously as a prop